### PR TITLE
Align importer models and serializers with MVP

### PR DESCRIPTION
**Pull request by asmacdo**
closes #2818
re #2817
closes #2816

This commit aligns the Importer model and serializer with the MVP
document for Pulp 3. The model docstrings are redundant with the
serializer help_text. The help_text will eventually become the
documentation, so removing using it exclusively will be the best way to
keep the code, comments, and documentation up to date.

### CODE REVIEW

File: platform/pulpcore/app/models/repository.py [hunk 1]
```diff
@@ -125,9 +103,16 @@ class Importer(ContentAdaptor):
     ON_DEMAND = 'on_demand'
     BACKGROUND = 'background'
     DOWNLOAD_POLICIES = (
-        (IMMEDIATE, 'Download Immediately'),
-        (ON_DEMAND, 'Download On Demand'),
-        (BACKGROUND, 'Download In Background'))
+        (IMMEDIATE, 'Sync task is not complete until downloading is finished.'),
+        (ON_DEMAND, 'Sync task updates metadata but does not download content until requested.'),
+        (BACKGROUND, 'Sync task updates metadata and completes. Downloads complete in background.'))
```
Comment (jortel) [inline platform/pulpcore/app/models/repository.py]: Suggest something more like:

IMMEDIATE: _Update the repository content and ensures all artifacts are downloaded._

ON_DEMAND: _Update the repository content but no artifacts are downloaded._

BACKGROUND: _Update the repository content and downloads artifacts in the background._

The _tasking_ seems irrelevant to the definition.

File: platform/pulpcore/app/models/repository.py [hunk 2]
```diff
@@ -95,28 +95,6 @@ def natural_key(self):
...
-        proxy_url (models.TextField): The optional proxy URL.
-            Format: scheme://user:password@host:port
-        basic_auth_user (models.TextField): The user used in HTTP basic authentication.
-        basic_auth_password (models.TextField): The password used in HTTP basic authentication.
-        download_policy (models.TextField): The policy for downloading content.
-        last_sync (models.DatetimeField): When the last successful synchronization occurred.
-
-    Relations:
-
-        scratchpad (GenericKeyValueRelation): Arbitrary information stashed by the importer.
```
Comment (jortel) [inline platform/pulpcore/app/models/repository.py]: I'm not sold on this change.  Also, it seems outside the scope of this issue.

File: platform/pulpcore/app/models/repository.py [hunk 3]
```diff
@@ -125,9 +103,16 @@ class Importer(ContentAdaptor):
...
+        (IMMEDIATE, 'Sync task is not complete until downloading is finished.'),
+        (ON_DEMAND, 'Sync task updates metadata but does not download content until requested.'),
+        (BACKGROUND, 'Sync task updates metadata and completes. Downloads complete in background.'))
+
+    # Sync Modes
+    ADDITIVE = 'additive'
+    MIRROR = 'mirror'
+    SYNC_MODES = (
+        (ADDITIVE, 'Add new content, keep content that was removed upstream.'),
+        (MIRROR, 'Add new content, remove content was removed upstream.'))
```
Comment (jortel) [inline platform/pulpcore/app/models/repository.py]: Suggest something like:

ADDITIVE:  _Add new content._

MIRROR: _Add new content and remove content that is no longer in the remote repository._

For additive, some content could still be removed from the repository in cases such as RPM only keeping X versions of RPMs.

File: platform/pulpcore/app/serializers/repository.py [hunk 4]
```diff
@@ -67,25 +67,27 @@ class ImporterSerializer(MasterModelSerializer, NestedHyperlinkedModelSerializer
...
         required=False,
     )
-
+    download_policy = serializers.ChoiceField(
+        help_text='The policy for downloading content.',
+        allow_blank=False,
+        choices=models.Importer.DOWNLOAD_POLICIES,
+    )
+    sync_mode = serializers.ChoiceField(
+        help_text='How the importer should sync from the upstream repository.',
```
Comment (jortel) [inline platform/pulpcore/app/serializers/repository.py]: I quess I lost the battle on pulp-dev for _sync_ vs _synchronize_ but really don't see value in using this abbreviation.  Is it really that long a word to type?

File: platform/pulpcore/app/serializers/repository.py [hunk 5]
```diff
@@ -67,25 +67,27 @@ class ImporterSerializer(MasterModelSerializer, NestedHyperlinkedModelSerializer
...
+        allow_blank=False,
+        choices=models.Importer.SYNC_MODES,
+    )
     validate = serializers.BooleanField(
-        help_text='Whether to validate imported content.',
+        help_text='If True, the plugin will validate imported content.',
```
Comment (jortel) [inline platform/pulpcore/app/serializers/repository.py]: To be clear, the content is not validated, the downloaded artifacts are.

File: platform/pulpcore/app/serializers/repository.py [hunk 6]
```diff
@@ -67,25 +67,27 @@ class ImporterSerializer(MasterModelSerializer, NestedHyperlinkedModelSerializer
...
         required=False,
     )
-
     ssl_ca_certificate = FileField(
         help_text='A PEM encoded CA certificate used to validate the server '
                   'certificate presented by the external source.',
```
Comment (jortel) [inline platform/pulpcore/app/serializers/repository.py]: Suggest removing: _presented by the external source_

File: platform/pulpcore/app/serializers/repository.py [hunk 7]
```diff
@@ -103,42 +105,41 @@ class ImporterSerializer(MasterModelSerializer, NestedHyperlinkedModelSerializer
...
-        help_text='The policy for downloading content.',
-        allow_blank=False,
-        choices=models.Importer.DOWNLOAD_POLICIES,
-    )
     last_sync = serializers.DateTimeField(
         help_text='Timestamp of the most recent successful sync.',
         read_only=True
     )
+    last_updated = serializers.DateTimeField(
+        help_text='Timestamp of the most recent update of the importer configuration.',
```
Comment (jortel) [inline platform/pulpcore/app/serializers/repository.py]: The importer fields are the configuration.  Suggest removing: _configuration_.

File: platform/pulpcore/app/serializers/repository.py [hunk 8]
```diff
@@ -103,42 +105,41 @@ class ImporterSerializer(MasterModelSerializer, NestedHyperlinkedModelSerializer
...
 
     class Meta:
         abstract = True
         model = models.Importer
         fields = MasterModelSerializer.Meta.fields + (
-            'name', 'last_updated', 'feed_url', 'validate', 'ssl_ca_certificate',
+            'name', 'feed_url', 'download_policy', 'sync_mode', 'validate', 'ssl_ca_certificate',
             'ssl_client_certificate', 'ssl_client_key', 'ssl_validation', 'proxy_url',
-            'basic_auth_user', 'basic_auth_password', 'download_policy', 'last_sync', 'repository',
+            'username', 'password', 'last_sync', 'last_updated', 'repository',
```
Comment (jortel) [inline platform/pulpcore/app/serializers/repository.py]: Large paragraph style lists are, IMHO, less readable.  consider:
```
...Meta.fields + (
    'name',
    'feed_url',
    ...
)
```

Also, easier to keep alphabetized (if so inclined).

Sorry, in the mood to be pedantic.

**Comment by jortel** [comment]
@asmacdo: These changes will likely break the FileImporter.  Please ensure a file plugin PR can be merged at the same time.

File: platform/pulpcore/app/serializers/repository.py (hunk 4 above)
Comment (asmacdo) [reply platform/pulpcore/app/serializers/repository.py]: Sync vs synchronize is often the difference between 1 and 2 lines. Obviously, this is a minor concern, but single lines are much easier to scan visually, and I think it has a noticeable impact on readability. As it turns out, sync is actually a real word, [informal]. http://www.dictionary.com/browse/sync

File: platform/pulpcore/app/models/repository.py [hunk 9]
```diff
@@ -97,23 +97,24 @@ class Importer(ContentAdaptor):
     Fields:
 
         feed_url (models.TextField): The URL of an external content source.
-        validate (models.BooleanField): Validate the imported context.
+        validate (models.BooleanField): If True, the plugin will validate imported files.
```
Comment (jortel) [inline platform/pulpcore/app/models/repository.py]: s/imported/downloaded/  might be clearer.

File: platform/pulpcore/app/models/repository.py [hunk 10]
```diff
@@ -146,16 +154,15 @@ class Importer(ContentAdaptor):
...
-
-    basic_auth_user = models.TextField(blank=True)
-    basic_auth_password = models.TextField(blank=True)
+    username = models.TextField(blank=True)
+    password = models.TextField(blank=True)
 
     download_policy = models.TextField(choices=DOWNLOAD_POLICIES)
-    last_sync = models.DateTimeField(blank=True, null=True)
+    sync_mode = models.TextField(choices=SYNC_MODES)
+    last_synced = models.DateTimeField(blank=True, null=True)
```
Comment (jortel) [inline platform/pulpcore/app/models/repository.py]: Is this read only?

Comment (jortel) [inline platform/pulpcore/app/models/repository.py]: Looks like the task is not updating last_synced.  I'll write an issue.

File: platform/pulpcore/app/serializers/repository.py [hunk 11]
```diff
@@ -74,25 +74,27 @@ class ImporterSerializer(MasterModelSerializer, NestedHyperlinkedModelSerializer
...
+        choices=models.Importer.DOWNLOAD_POLICIES,
+    )
+    sync_mode = serializers.ChoiceField(
+        help_text='How the importer should sync from the upstream repository.',
+        allow_blank=False,
+        choices=models.Importer.SYNC_MODES,
+    )
     validate = serializers.BooleanField(
-        help_text='Whether to validate imported content.',
+        help_text='If True, the plugin will validate imported artifacts.',
```
Comment (jortel) [inline platform/pulpcore/app/serializers/repository.py]: s/imported/downloaded/

File: platform/pulpcore/app/models/repository.py (hunk 9 above)
Comment (asmacdo) [reply platform/pulpcore/app/models/repository.py]: Is this also used for uploads?

Comment (jortel) [reply platform/pulpcore/app/models/repository.py]: dunno

Comment (asmacdo) [reply platform/pulpcore/app/models/repository.py]: @dkliban, I am going to leave this for today and merge. If uploads don't use this field, maybe we should specify that it is only for downloads.
